### ISSUE TITLE
JUnit reporter using too much memory

### ISSUE THREAD

**Issue by davnat** [catchorg/Catch2#1546]
I'm having issues with how much memory is used by the test binary generated by Catch when using the JUnit reporter.
Here I'm using Valgrind to see how much memory is used by the program.
Please note that the memory reported by Valgrind is not necessarily all in use by the program at a certain point in time, nevertheless the numbers are quite impressive to me.

```xml
$ valgrind --leak-check=full ./build/testbin
==20597== Memcheck, a memory error detector
==20597== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==20597== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==20597== Command: ./build/testbin
==20597== 
===============================================================================
All tests passed (4456282 assertions in 51 test cases)

==20597== 
==20597== HEAP SUMMARY:
==20597==     in use at exit: 0 bytes in 0 blocks
==20597==   total heap usage: 293,728 allocs, 293,728 frees, 39,396,626 bytes allocated
==20597== 
==20597== All heap blocks were freed -- no leaks are possible
==20597== 
==20597== For counts of detected and suppressed errors, rerun with: -v
==20597== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
This is ~40 MB of RAM used with the default reporter, with something less that 300k allocations.
```xml
$ valgrind --leak-check=full ./build/testbin -r junit
==20648== Memcheck, a memory error detector
==20648== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==20648== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==20648== Command: ./build/testbin -r junit
==20648== 
<?xml version="1.0" encoding="UTF-8"?>
==20648== Warning: set address range perms: large range [0x27513040, 0x46513040) (undefined)
<testsuites>
  <testsuite name="testbin" errors="0" failures="0" tests="4458927" hostname="tbd" time="224.856" timestamp="2019-02-18T16:15:56Z">

(~110 testcase entries)

    <system-out/>
    <system-err/>
  </testsuite>
</testsuites>
==20648== Warning: set address range perms: large range [0x27513028, 0x46513058) (noaccess)
==20648== 
==20648== HEAP SUMMARY:
==20648==     in use at exit: 0 bytes in 0 blocks
==20648==   total heap usage: 4,315,909 allocs, 4,315,909 frees, 3,212,886,130 bytes allocated
==20648== 
==20648== All heap blocks were freed -- no leaks are possible
==20648== 
==20648== For counts of detected and suppressed errors, rerun with: -v
==20648== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

This makes a total of 3.2 GB of RAM used, with 4.3 millions of allocations!

This makes me trouble, because I have to run the test binary on an Android emulator, and this forces me to run an emulator with 4 GB of RAM or so.

Is there something I could do to make the JUnit reporter a bit less memory hungry?

Thanks, cheers.

**Platform information:**
 - OS: **Ubuntu 18.04**
 - Compiler+version: **GCC v7.3.0**
 - Catch version: **v2.6.1**

**Comment by horenmar** [comment]
You can

1) Not use the JUnit reporter
2) Run the tests in parts -- a trick we use at my job is to use the `-#` flag and essentially run tests from each TU on its own. We do it for parallelism rather than decreasing memory consumption, but it should work for you as well.

The reason why the JUnit reporter takes so much RAM is that the format essentially requires the reporter to first run all tests, hold on all the relevant data (original expressions, expression stringifications, location info, etc) and only after all the tests have finished, it can start printing out the XML.

**Comment by davnat** [comment]
IMHO a "wontfix" would be more appropriate: I fail to see how this could be "resolved".

**Comment by horenmar** [comment]
🤷‍♂️  Resolved is a temporary label that says "close this issue later unless there is a relevant followup".
